Assets/Scripts/Encounters/ShopEncounter.cs

```diff
@@ -101,47 +102,71 @@ public override void build(List<Companion> companionList, EncounterConstantsSO c
-        int commonCardPercentage = shopData.commonCardPercentage;
-        int uncommonCardPercentage = shopData.uncommonCardPercentage;
-        int rareCardPercentage = shopData.rareCardPercentage;
+    private void generateCards(List<Companion> companionList) {
+        Debug.Log("Companion List count: " + companionList.Count);
+        //determine which companion to spawn a card from, remove them from the set
+        //move companion types to a hashSet
+        HashSet<CompanionTypeSO> companionTypes = new HashSet<CompanionTypeSO>();
+
+        foreach (Companion companion in companionList) {
+            companionTypes.Add(companion.companionType);
+        }
+
+        Debug.Log("Companion types: " + companionTypes.Count);
+
+        foreach (CompanionTypeSO companionType in companionTypes) {
+            //pick a random card based on random algorithm
+            CardPoolSO currentCardPool = companionType.cardPool;
+            int totalPercentage = currentCardPool.getTotalCardPercentage();
+            int commonCardPercentage = currentCardPool.commonCardPercentage;
+            int uncommonCardPercentage = currentCardPool.uncommonCardPercentage;
+            int rareCardPercentage = currentCardPool.rareCardPercentage;
+
 
-        for (int i = 0; i < numCardsToGenerate; i++) {
             // Determine what the card pool is for this single card being generated
-            List<CardType> cardPool = new List<CardType>();
+            SerializableHashSet<CardType> cardSet = new SerializableHashSet<CardType>();
             int cardPrice = 0;
             int randomNumber = Random.Range(0, totalPercentage); // min inclusive, max exclusive
             if (randomNumber < commonCardPercentage) {
-                cardPool = shopData.commonCards;
-                cardPrice = shopData.commonCardPrice;
+                cardSet = currentCardPool.commonCards;
+                cardPrice = currentCardPool.commonCardPrice;
             } else if (randomNumber < commonCardPercentage + uncommonCardPercentage) {
-                cardPool = shopData.uncommonCards;
-                cardPrice = shopData.uncommonCardPrice;
+                cardSet = currentCardPool.uncommonCards;
+                cardPrice = currentCardPool.uncommonCardPrice;
             } else {
-                cardPool = shopData.rareCards;
-                cardPrice = shopData.rareCardPrice;
+                cardSet = currentCardPool.rareCards;
+                cardPrice = currentCardPool.rareCardPrice;
             }
 
             // Pick a card from the pool and add it to the shop's cards
-            int cardNumber = Random.Range(0, cardPool.Count);
-            cardsInShop.Add(
-                new CardInShopWithPrice(cardPool[cardNumber], cardPrice));
+            int cardNumber = Random.Range(0, cardSet.Count);
+
+            //super cool and efficent random selection from a hashSet(it has to iterate through the collection, there is no index in a hashset)
```

Comment:
This is interesting and got me thinking about whether we should be using a hashset here.. realistically these cardpools are never going to be larger than ~200 cards each, are we doing this for reasons other than performance?

Reply:
This was for the .Contains method performance that you commented on below